googletest/src/gtest.cc

```diff
@@ -723,60 +723,91 @@ static bool PatternMatchesString(const std::string& name_str,
   return true;
 }
 
-bool UnitTestOptions::MatchesFilter(const std::string& name_str,
-                                    const char* filter) {
-  // The filter is a list of patterns separated by colons (:).
-  const char* pattern = filter;
-  while (true) {
-    // Find the bounds of this pattern.
-    const char* const next_sep = strchr(pattern, ':');
-    const char* const pattern_end =
-        next_sep != nullptr ? next_sep : pattern + strlen(pattern);
-
-    // Check if this pattern matches name_str.
-    if (PatternMatchesString(name_str, pattern, pattern_end)) {
-      return true;
+class Filter {
+  std::vector<std::string> patterns_;
+
+ public:
+  // Constructs a filter form a string of patterns separated by `:`.
+  explicit Filter(const std::string& filter) {
+    if (filter.empty()) return;
+
+    auto pattern_start = filter.begin();
+    while (true) {
+      const auto delimiter = std::find(pattern_start, filter.end(), ':');
+      patterns_.emplace_back(pattern_start, delimiter);
+
+      if (delimiter == filter.end())
+        break;
+      pattern_start = std::next(delimiter);
     }
+  }
 
-    // Give up on this pattern. However, if we found a pattern separator (:),
-    // advance to the next pattern (skipping over the separator) and restart.
-    if (next_sep == nullptr) {
-      return false;
+  // Returns true if and only if name matches at least one of the patterns in
+  // the filter.
+  bool MatchesName(const std::string& name) const {
+    const auto pattern_matches_name = [&name](const std::string& pattern) {
+      return PatternMatchesString(name, pattern.c_str(),
+                                  pattern.c_str() + pattern.size());
+    };
+    return std::any_of(patterns_.begin(), patterns_.end(),
+                       pattern_matches_name);
+  }
+};
+
+class PositiveAndNegativeFilter {
+  Filter positive_filter_;
+  Filter negative_filter_;
+
+  static std::pair<std::string, std::string>
+  BreakIntoPositiveAndNegativeFilters(const std::string& filter) {
+    const auto dash_pos = filter.find('-');
```

Comment:
Can we reuse the `SplitString` function defined in this file?
https://github.com/google/googletest/blob/3e0e32ba300ce8afe695ad3ba7e81b21b7cf237a/googletest/src/gtest.cc#L1156

Reply:
resolved.